byway/src/main/webapp/recommendations.js

```diff
@@ -0,0 +1,254 @@
+ * @param {google.maps.LatLng} end coordinate of route's endpoint
+ */
+function addDestinations(start, end) {
+  destinations.add(start);
+  destinations.add(end);
+  regions.push(...destinations);
+}
+
+/**
+ * Creates a route between two points and loads onto the map.
+ * Finds points along the path to load the regions Array.
+ * Loads recommendations centered around these points.
+ * @param {DirectionsService} directionsService finds directions
+ * @param {DirectionsRenderer} directionsRenderer renders the route
+ * @param {LatLng} start starting point location
+ * @param {LatLng} end ending point location
+ */
+function calcRoute(directionsService, directionsRenderer, start, end) {
+  const request = {
+      origin:  start,
+      destination: end,
+      travelMode: 'DRIVING'
+  };
+  directionsService.route(request, function(result, status) {
+    if (status == 'OK') {
+      directionsRenderer.setDirections(result);
+      findRegions(result);
+      loadRecommendations();
+    } else {
+      alert("Could not calculate route due to: " + status);
+    }
+  });
+}
+
+/**
+ * Goes through steps of every leg along route to find the average location
+ * between a step's start and end location. Store these points
+ * in the regions Array.
+ * @param {DirectionsResult} directionResult contains directions
+ * for the route made.
+ */
+function findRegions(directionResult) {
+  const myRoute = directionResult.routes[0];
+  for(leg of myRoute.legs) {
+    for(step of leg.steps) {
+      const avgLat = (step.start_location.lat() + step.end_location.lat()) / 2;
+      const avgLng = (step.start_location.lng() + step.end_location.lng()) / 2;
+      const avgLoc = {lat: avgLat, lng: avgLng};
+      if(step.distance.value > MIN_DISTANCE) {
+        regions.push(avgLoc);
+      }
+    }
+  }
+}
+
+/**
+ * Set a timeout to delay the browser.
+ * @param {Number} delayMs number of milliseconds
+ * @return empty promise after delayMs milliseconds passed
+ */
+function delayPromise(delayMs) {
+  return new Promise(resolve => setTimeout(resolve, delayMs));
+}
+
+/**
+ * Go through a user's interests and search for places
+ * fitting those interests. Search around the regions
+ * previously found using textSearch. Prioritize finding
+ * past results loaded through sessionStorage to avoid calling
+ * textSearch repeatedly from the PlacesService.
+ */
+async function loadRecommendations() {
+  for(interest of interests) {
+    for(region of regions) {
+      const request = {
+        location: region,
+        radius: RADIUS,
+        query: interest
+      }
+      const recommendationsSaved = sessionStorage.getItem(JSON.stringify(request));
+      if(recommendationsSaved !== null) {
+        addRecommendations(request, JSON.parse(recommendationsSaved));
+      } else{
+        // Prevent hitting the query limit from the maps API
+        // by setting a timeout in delayPromise
+        await delayPromise(250);
```

Comment:
Does this actually work the way you intend? It looks like it is just setting a timer for 250 ms before continuing execution. Is the idea wait for multiple calls to `loadRecommendations()`

Reply:
This timeout is an intermediate step when making a new textSearch request. When doing `await findPlacesWithTextSearch()`, it waits for the function to return a new promise guaranteed to be completed before it continues executing. Then when it finishes adding them, it continues in the for loop to find the new results for a different interest/region.

If I remove this `await delayPromise()` line, the code awaits like normal when making a new call to `findPlacesWithTextSearch()`. The issue I saw, however, was when refreshing, as the await seems to not "wait" as much with results that were already generated. Not sure how this works, I assumed it was cached or something similar. But, it hit the query limit if too many results were found before refreshing, since loading them after the refresh didnt work the same way for previously-loaded entries.

Reply:
@landondare have a look at the constructor for `Promise` works.